File: resources/ios/ActivityIndicatorRenderer.swift
import Combine
import SwiftUI

struct ActivityIndicatorRendererConfiguration: Equatable {
    let size: ActivityIndicatorSize
    let accessibilityLabel: String

    init(node: NativeUINode) {
        size = ActivityIndicatorSize(
            rawValue: node.props.getString("size", default: "md")
        ) ?? .medium
        accessibilityLabel = node.props.getString("a11y_label")
    }
}

struct ActivityIndicatorAnnouncementState {
    private var hasAnnounced = false

    mutating func consume(label: String) -> String? {
        guard !hasAnnounced, !label.isEmpty else { return nil }

        hasAnnounced = true
        return label
    }
}

struct ActivityIndicatorRendererState {
    private(set) var configuration: ActivityIndicatorRendererConfiguration
    private let nodeID: Int
    private var announcement = ActivityIndicatorAnnouncementState()

    init(node: NativeUINode) {
        nodeID = node.id
        configuration = ActivityIndicatorRendererConfiguration(node: node)
    }

    mutating func consumeAnnouncement() -> String? {
        announcement.consume(label: configuration.accessibilityLabel)
    }

    @discardableResult
    mutating func serverPublished(_ tree: NativeUITree) -> Bool {
        guard let node = activityIndicatorNode(id: nodeID, in: tree.root) else {
            return false
        }

        let published = ActivityIndicatorRendererConfiguration(node: node)
        guard published != configuration else { return false }

        configuration = published
        return true
    }
}

struct ActivityIndicatorRenderer: View {
    let node: NativeUINode

    @ObservedObject private var themeStore = NativeUITheme.shared
    @Environment(\.colorScheme) private var colorScheme
    @State private var state: ActivityIndicatorRendererState

    init(node: NativeUINode) {
        self.node = node
        _state = State(initialValue: ActivityIndicatorRendererState(node: node))
    }

    var body: some View {
        let configuration = state.configuration

        FirstlightActivityIndicatorControl(
            size: configuration.size,
            accessibilityLabel: configuration.accessibilityLabel,
            tint: themeStore.resolve(for: colorScheme).primary
        )
        .onAppear {
            if let label = state.consumeAnnouncement() {
                AccessibilityNotification.Announcement(label).post()
            }
        }
        .onReceive(NativeUIBridge.shared.$currentTree.compactMap { $0 }) { tree in
            state.serverPublished(tree)
        }
    }
}

private func activityIndicatorNode(id: Int, in node: NativeUINode) -> NativeUINode? {
    if node.id == id { return node }

    for child in node.children {
        if let found = activityIndicatorNode(id: id, in: child) {
            return found
        }
    }

    return nil
}

File: resources/android/ActivityIndicatorRenderer.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import com.nativephp.mobile.ui.nativerender.NativeUIBridge
import com.nativephp.mobile.ui.nativerender.NativeUINode
import com.nativephp.mobile.ui.nativerender.NativeUITree
import com.nativephp.plugins.native_ui.NativeUITheme

data class ActivityIndicatorRendererConfiguration(
    val size: ActivityIndicatorSize,
    val accessibilityLabel: String,
) {
    constructor(node: NativeUINode) : this(
        size = ActivityIndicatorSize.fromWire(node.props.getString("size", "md")),
        accessibilityLabel = node.props.getString("a11y_label"),
    )

    val isInteractive = false
}

class ActivityIndicatorRendererState(node: NativeUINode) {
    private val nodeId = node.id

    var configuration by mutableStateOf(ActivityIndicatorRendererConfiguration(node))
        private set

    fun serverPublished(tree: NativeUITree): Boolean {
        val node = tree.root.findActivityIndicatorNode(nodeId) ?: return false
        val published = ActivityIndicatorRendererConfiguration(node)
        if (published == configuration) return false

        configuration = published
        return true
    }
}

object FirstlightActivityIndicatorRenderer {
    @Composable
    fun Render(node: NativeUINode, modifier: Modifier) {
        val state = remember(node.id) { ActivityIndicatorRendererState(node) }
        val publishedTree = NativeUIBridge.currentTree.value
        LaunchedEffect(publishedTree) {
            publishedTree?.let(state::serverPublished)
        }

        val tokens = if (isSystemInDarkTheme()) NativeUITheme.dark else NativeUITheme.light
        val configuration = state.configuration

        FirstlightActivityIndicatorControl(
            size = configuration.size,
            accessibilityLabel = configuration.accessibilityLabel,
            color = tokens.primary,
            modifier = modifier,
        )
    }
}

private fun NativeUINode.findActivityIndicatorNode(id: Int): NativeUINode? {
    if (this.id == id) return this

    children.forEach { child ->
        child.findActivityIndicatorNode(id)?.let { return it }
    }

    return null
}
